### ISSUE TITLE
early error on ./opam_installer.sh: cannot see the error message

### ISSUE THREAD

**Issue by thranur** [ocaml/opam#2252]
I just installed Linux Mint on a virtual machine and tried to install opam using `opam_installer.sh`. I do have gcc, but some basic package must be missing, however the error message is not clear and the `.opam` directory was not created, so I cannot find out exactly what happened.

Here's the command log:

```
and@vbmint ~ $ ./opam_installer.sh opam
Downloading OPAM...
Initializing with compiler 4.02.1
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.


=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[default] synchronized from https://opam.ocaml.org

=-=- Installing compiler 4.02.1 -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[compiler.get] http://caml.inria.fr/pub/distrib/ocaml-4.02/ocaml-4.02.1.tar.gz downloaded
Now compiling OCaml. This may take a while, please bear with us...
[ERROR] Compiler build failed at "./configure -prefix /home/and/.opam/4.02.1 -with-debug-runtime":
        # opam-version 1.2.2
        # os           linux
        # command      ./configure -prefix /home/and/.opam/4.02.1 -with-debug-runtime
        # path         /home/and/.opam/4.02.1/build/ocaml
        # exit-code    2
        # env-file     /home/and/.opam/log/log-18926-b1c0e4.env
        # stdout-file  /home/and/.opam/log/log-18926-b1c0e4.out
        # stderr-file  /home/and/.opam/log/log-18926-b1c0e4.err
        ### stdout ###
        # Configuring for host x86_64-unknown-linux-gnu ...
        # Configuring for target x86_64-unknown-linux-gnu ...
        # Using compiler gcc.
        # [ERROR!]Unable to compile the test program.
        #  Make sure the C compiler gcc -O  is properly installed.

[ERROR] Initialisation failed
OpamGlobals.Exit(66)
and@vbmint ~ $ cat /home/and/.opam/log/log-18926-b1c0e4.err
cat: /home/and/.opam/log/log-18926-b1c0e4.err: No such file or directory
and@vbmint ~ $ ls /home/and/.opam
ls: cannot access /home/and/.opam: No such file or directory
and@vbmint ~ $ gcc -O
gcc: fatal error: no input files
compilation terminated.
```

**Comment by AltGr** [comment]
It seems opam itself has been correctly installed, but initialisation failed. You may try first installing the ocaml version that is available in Mint, which should get you the proper dependencies.

Otherwise, running `opam init --verbose` should give you more details on what's happening.

**Comment by thranur** [comment]
My intention was to bootstrap OPAM + OCaml from a clean machine, without using the distribution package, in order to see exactly what is needed in a new machine, so that I would try to make a foolproof script to install opam on some machines (think "first-year students who barely used Linux at all"). Also, I prefer to avoid at all costs having a distribution-based OCaml and another via OPAM, it always leads me to mixing up both with terrible consequences.

If this use case ("OPAM clean install from zero") could be supported by OPAM, then I'll investigate which tools are missing. For instance, I purposely did not install `build-essential`, which is probably one of the causes.

My point was that (1) the error message was not exactly precise, since `gcc` is installed, and (2) it seems that the initialization code does not have time to write the error files when the error happens very early in the installation, or it deletes them because of the errors, so I'd like to help improve on that.

I tried `./opam init --verbose` on the directory where opam-installer put the downloaded `opam` binary, but again I did not get much information (extract below).

I'll try looking into the source to see what's happening, if the directories mentioned in the error message (env-file, stdout-file, etc.) never got created in the first place, or if they were deleted afterwards.

```
and@vbmint ~/opam $ ./opam init --verbose --comp=4.02.2
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.


=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[default] synchronized from https://opam.ocaml.org

=-=- Installing compiler 4.02.2 -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[compiler.get] http://caml.inria.fr/pub/distrib/ocaml-4.02/ocaml-4.02.2.tar.gz downloaded
Now compiling OCaml. This may take a while, please bear with us...
+ ./configure "-prefix" "/home/and/.opam/4.02.2" "-with-debug-runtime" (CWD=/home/and/.opam/4.02.2/build/ocaml)
- Configuring for host x86_64-unknown-linux-gnu ...
- Configuring for target x86_64-unknown-linux-gnu ...
- Using compiler gcc.
- [ERROR!] Unable to compile the test program.
-  Make sure the C compiler gcc -O  is properly installed.
[ERROR] Compiler build failed at "./configure -prefix /home/and/.opam/4.02.2 -with-debug-runtime":
        # opam-version 1.2.2
        # os           linux
        # command      ./configure -prefix /home/and/.opam/4.02.2 -with-debug-runtime
        # path         /home/and/.opam/4.02.2/build/ocaml
        # exit-code    2
        # env-file     /home/and/.opam/log/log-3612-06f473.env
        # stdout-file  /home/and/.opam/log/log-3612-06f473.out
        # stderr-file  /home/and/.opam/log/log-3612-06f473.err
        ### stdout ###
        # Configuring for host x86_64-unknown-linux-gnu ...
        # Configuring for target x86_64-unknown-linux-gnu ...
        # Using compiler gcc.
        # [ERROR!] Unable to compile the test program.
        #  Make sure the C compiler gcc -O  is properly installed.

[ERROR] Initialisation failed
OpamGlobals.Exit(66)
'./opam init --verbose --comp=4.02.2' failed.
```

**Comment by AltGr** [comment]
> Also, I prefer to avoid at all costs having a distribution-based OCaml and another via OPAM, it always leads me to mixing up both with terrible consequences.

There really shouldn't be a problem with opam, as this is a very common use-case: `opam switch` can be used to handle multiple compilers, and the special `system` switch will use the one from the distribution. But that's besides the question.

> I'll try looking into the source to see what's happening, if the directories mentioned in the error message (env-file, stdout-file, etc.) never got created in the first place, or if they were deleted afterwards.

With `--debug`, they shouldn't get removed.

As for the prerequisites:
- requirements for building OCaml as per https://github.com/ocaml/ocaml/blob/trunk/INSTALL . That's probably just GNU make and gcc in your case. You may want to just get the OCaml source and run the `./configure` command to see what's going on.
- `apt-cache show opam` lists: `Depends: libbz2-1.0, libc6 (>= 2.14), zlib1g (>= 1:1.1.4), opam-docs (= 1.2.0-1), build-essential, wget | curl`. You'll need one of `wget` or `curl`, `tar`, and although opam should work you probably won't build many packages without `build-essential`.
- although that may pass for simple use, an [external solver](http://opam.ocaml.org/doc/Install.html#ExternalSolvers) is recommended.

**Comment by thranur** [comment]
Thank you, I didn't know `--debug` allowed the files to remain (`opam help` only mentioned "Print debug message to stderr.", although `configure` does mention that `--debug` keeps temporary files). That said, they don't contain as much information as I expected...

Following your second suggestion, using the OPAM sources I was able to see the actual ./configure error message:

```
configure:2127: checking whether the C compiler works
configure:2149: gcc    conftest.c  >&5
/usr/bin/ld: cannot find crt1.o: No such file or directory
/usr/bin/ld: cannot find crti.o: No such file or directory
/usr/bin/ld: cannot find -lc
/usr/bin/ld: cannot find crtn.o: No such file or directory
collect2: error: ld returned 1 exit status
(...)
configure:2196: error: in `/home/and/git/opam':
configure:2198: error: C compiler cannot create executables
```

The minimal solution is to install `libc6-dev`. Not sure it's worth trying to give a more specific error message in this case, since having a gcc which cannot even compile a hello world is not very common.

Finally, a small (but possibly useless) remark: when using `--debug`, files such as `.opam/config` are not erased even if the initialization failed, which means that trying to run `opam init` again afterwards leads to "OPAM is already initialized", even if the `bin` directory of the OCaml compiler is empty because of the previous errors. Again, such an improbable case that it does not need fixing in my opinion.

Overall, I found out the extra condition to test in my foolproof script, but it's so improbable that probably does not warrant fixing. Thanks for the tips and sorry for the noise.
